core/src/ledger/masp_utils.rs

```diff
@@ -0,0 +1,113 @@
+use crate::types::address::MASP;
+use crate::types::hash::Hash;
+use crate::types::storage::{BlockHeight, Epoch, Key, KeySeg, TxIndex};
+use crate::types::token::{
+    Transfer, HEAD_TX_KEY, MASP_NOTE_COMMITMENT_TREE_KEY, MASP_NULLIFIERS_KEY,
+    PIN_KEY_PREFIX, TX_KEY_PREFIX,
+};
+
+// Writes the nullifiers of the provided masp transaction to storage
+fn reveal_nullifiers(
+    ctx: &mut impl StorageWrite,
+    transaction: &Transaction,
+) -> Result<()> {
+    for description in transaction
+        .sapling_bundle()
+        .map_or(&vec![], |description| &description.shielded_spends)
+    {
+        let nullifier_key = Key::from(MASP.to_db_key())
+            .push(&MASP_NULLIFIERS_KEY.to_owned())
+            .expect("Cannot obtain a storage key")
+            .push(&Hash(description.nullifier.0))
+            .expect("Cannot obtain a storage key");
+        ctx.write(&nullifier_key, ())?;
+    }
+
+    Ok(())
+}
+
+/// Appends the note commitments of the provided transaction to the merkle tree
+/// and updates the anchor
+/// NOTE: this function is public as a temporary workaround because of an issue
+/// when running this function in WASM
+pub fn update_note_commitment_tree(
+    ctx: &mut (impl StorageRead + StorageWrite),
+    transaction: &Transaction,
+) -> Result<()> {
+    if let Some(bundle) = transaction.sapling_bundle() {
+        if !bundle.shielded_outputs.is_empty() {
+            let tree_key = Key::from(MASP.to_db_key())
+                .push(&MASP_NOTE_COMMITMENT_TREE_KEY.to_owned())
+                .expect("Cannot obtain a storage key");
+            let mut commitment_tree: CommitmentTree<Node> =
+                ctx.read(&tree_key)?.ok_or(Error::SimpleMessage(
+                    "Missing note commitment tree in storage",
+                ))?;
+
+            for description in &bundle.shielded_outputs {
+                // Add cmu to the merkle tree
+                commitment_tree
+                    .append(Node::from_scalar(description.cmu))
+                    .map_err(|_| {
+                        Error::SimpleMessage("Note commitment tree is full")
+                    })?;
+            }
+
+            ctx.write(&tree_key, commitment_tree)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Handle a MASP transaction.
+pub fn handle_masp_tx(
+    ctx: &mut (impl StorageRead + StorageWrite),
+    transfer: &Transfer,
+    shielded: &Transaction,
+) -> Result<()> {
+    let masp_addr = MASP;
+    let head_tx_key = Key::from(masp_addr.to_db_key())
+        .push(&HEAD_TX_KEY.to_owned())
+        .expect("Cannot obtain a storage key");
+    let current_tx_idx: u64 =
+        ctx.read(&head_tx_key).unwrap_or(None).unwrap_or(0);
+    let current_tx_key = Key::from(masp_addr.to_db_key())
+        .push(&(TX_KEY_PREFIX.to_owned() + &current_tx_idx.to_string()))
+        .expect("Cannot obtain a storage key");
+    // Save the Transfer object and its location within the blockchain
```

Comment:
Do we really need to store the entire transaction data? Seems wasteful. Also, can't this be queried from Tendermint?

Reply:
We don't need this data in protocol but we need it for the clients to scan transactions and generate a valid internal state to produce future transactions. I agree with you that this could bloat the storage pretty quick (together with the nullifier set), but I think there are a few concerns with querying this data from comet:

- I'm not sure we can assume that the blocks will be always available (in case of a chain restart for example if we decided to prune history)
- Requesting blocks from full nodes could take a long time/bandwidth
- The time to process the blocks filtering out only masp transactions could be long

Still we might prefer this solution over writing to storage. Wdyt? Alternatively we could think about a way to prune transactions from storage when they are not useful anymore (e.g. all the output descriptions have been spent)

Reply:
We should be able to use the indexer to pre-filter for MASP transaction data so that clients need only download and scan that. It's true that we need to keep around data after chain upgrades, but this doesn't seem too difficult.